File: Java/src/com/dbc/algorithm/Pointer/MinimumWindowSubstring.java
package com.dbc.algorithm.Pointer;

import java.util.HashMap;
import java.util.Map;

public class MinimumWindowSubstring {
    public static String minWindow(String s, String t) {
        Map<Character, Integer> remain = new HashMap<>();
        Map<Character, Integer> store = new HashMap<>();

        for (int i = 0; i < t.length(); i++) {
            char alpha = t.charAt(i);
            store.put(alpha, store.getOrDefault(alpha, 0) + 1);
            remain.put(alpha, 0);
        }

        int left = -1, right = -1, count = 0, min_length = 100000;
        int i = 0, j = 0, length = t.length();
        boolean flag = true;
        while (i <= j && j < s.length()) {
            char iAlpha = s.charAt(i);
            char jAlpha = s.charAt(j);
            int temp = remain.getOrDefault(jAlpha, -1);

            if (flag && temp != -1) {
                remain.put(jAlpha, temp + 1);
                if (temp + 1 <= store.get(jAlpha))
                    count++;
            }

            if (count == length) {
                if (min_length > j - i + 1) {
                    left = i;
                    right = j;
                    min_length = j - i + 1;
                }
                if (remain.getOrDefault(iAlpha, -1) != -1) {
                    remain.put(iAlpha, remain.get(iAlpha) - 1);
                    if (remain.get(iAlpha) < store.get(iAlpha))
                        count--;
                }
                i++;
            }

            if (count == length) {
                flag = false;
            } else {
                j++;
                flag = true;
            }
        }

        if (min_length == 100000)
            return "";
        else {
            return s.substring(left, right + 1);
        }
    }
}

File: Python/pointer/minimum_window_substring.py
def minWindow(s: str, t: str) -> str:
    length = len(t)
    left, right, min_length = -1, -1, 100000

    restore = {}
    remain = {}
    for alpha in t:
        restore[alpha] = 1 if restore.get(alpha, -1) == -1 else restore[alpha] + 1
        remain[alpha] = 0

    i, j, count = 0, 0, 0
    flag = True
    while i <= j and j < len(s):
        if flag and remain.get(s[j], -1) != -1:
            remain[s[j]] += 1
            if remain[s[j]] <= restore[s[j]]:
                count += 1
        if count == length:
            if min_length > j - i + 1:
                left, right = i, j
                min_length = j - i + 1
            if remain.get(s[i], -1) != -1:
                remain[s[i]] -= 1
                if remain[s[i]] < restore[s[i]]:
                    count -= 1
            
            i += 1

        if count == length:
            flag = False
        else:
            flag = True
            j += 1

    if min_length == 100000:
        return ""
    else:
        return s[left:right+1]


if __name__ == "__main__":
    print(minWindow(s = "ADOBECODEBANC", t = "ABC"))
